Stored certificates in filesystem with LittleFS

SamGeens opened the pull request:
Hi

I started using your libary a while ago, and I love it! I succesfully finished my project, with a few small changes to your library and I'm willing to share those improvements. I hope you like pull requests because I have a few more coming your way. 

The first change I made is an improvement you already had on your TODO list:
1. Enabling LittleFS and storing certificates there, the idea is to validate TLS connection against the list of common CA certificates stored locally. No CRL though.

I implemented this the best I could, and documented everything you and other users need to know in the README.md. It seems that my code formatter prettified the code. I also changed the command order (without changing anything to the commands), to have the same order as [The Espressif AT Command Set](https://docs.espressif.com/projects/esp-at/en/latest/AT_Command_Set/index.html) so it is easy to see which command are not (yet) implemented.

Take your time to review this pull request and let me know what you think. I hope you like it :)

JiriBilek commented:
Thanks, you made a lot of work!
I will review the changes and let you know.

jandrassy commented:
@SamGeens did you see this?
https://github.com/JiriBilek/ESP_ATMod/pull/4/files

SamGeens commented:
@jandrassy Do you mean something specific? But yes I checked the files

PR should use squash commit for sure!

JiriBilek commented:
I quickly reviewed the code and it looks good. But please next time refrain from reformatting the original code however you like the formatting or not. It introduces many changes that I have to check.
I'll test it on an ESP device and let you now.

SamGeens commented:
Yes ofcourse, my code formatter ran and I thought pushing it wouldn't harm anyone. Next time I won't push any new formatting.

SamGeens commented:
I discovered a small mistake with checking for commands which I fixed in my next PR. (Which will be summitted after this one is accepted and squashed so I don't keep stacking up all those commits).

JiriBilek commented:
Thanks, I've already checked all files with unimportant or obvious changes.
I was too busy today but tomorrow here should be raining so I expect I'll make some progress.

SamGeens commented:
In case of any trouble with exact match commands with '?' (which my new PR will fix). You can already try the updated findCommand method.

`
/*********************************************************************************************
 * Searches the input buffer for a command. Returns command code or CMD_ERROR
 */
commands_t findCommand(uint8_t *input, uint16_t inpLen)
{
	// Check the buffer invariants
	if (inpLen < 4 || input[0] != 'A' || input[1] != 'T' || input[inpLen - 2] != '\r' || input[inpLen - 1] != '\n')
		return CMD_ERROR;

	// AT command as a special case
	if (inpLen == 4)
		return CMD_AT;

	char *inputCmd = (char *)calloc(strlen((char *)input + 2) + 1, sizeof(char));
	strcpy(inputCmd, (char *)input + 2);

	// Check the command list
	for (unsigned int i = 0; i < sizeof(commandList) / sizeof(commandDef_t); ++i)
	{
		const char *cmd = commandList[i].text;
		String compareCmd = commandList[i].text;

		inputCmd[strcspn(inputCmd, "\r\n")] = 0;

		// Split input command on \r or \n or any number
		inputCmd = strtok(inputCmd, "?=0123456789");

		// Remove trailing ? for comparison
		compareCmd.remove(compareCmd.lastIndexOf('?'));
		if (!strcmp(compareCmd.c_str(), inputCmd))
		{
			// We have a command
			switch (commandList[i].mode)
			{
			case MODE_NO_CHECKING:
				break;

			case MODE_EXACT_MATCH:
				if (inpLen != strlen(cmd) + 4) // Check exact length
					return CMD_ERROR;
				break;

			case MODE_QUERY_SET:
			{
				char c = input[strlen(cmd) + 2];
				if (c != '=' && c != '?')
					return CMD_ERROR;
				if (c == '?' && inpLen != strlen(cmd) + 5) // Check exact length
					return CMD_ERROR;
			}
			break;

			default:
				return CMD_ERROR;
			}

			return commandList[i].cmd;
		}
	}

	return CMD_ERROR;
}
`

JiriBilek commented on ESP_ATMod/command.cpp:
```diff
@@ -1784,263 +2093,196 @@ void cmd_AT_CIPRECVMODE()
+	if (inputBufferCnt == 15)
 	{
-		dnsConfig_t cfg;
+		PemCertificate = new char[MAX_PEM_CERT_LENGTH];
```
Should not be called if `(CAcert.getCount() >= maximumCertificates)`, otherwise it is not deleted and causes memory leaks.

SamGeens replied:
Fixed this one.

JiriBilek commented:
@SamGeens , I really appreciate your work but the fact that your changed the order of functions in command.cpp makes it nearly impossible to review and comment. Would it be much work for you to edit the file and restore the original order?
I understand the reason why you did it and we can do it again in the future but now it's a pain in the ass to compare the original file and the new one. Sorry :(

JiriBilek commented:
I'd like to know your opinion about the syntax of AT+CIPSSLCERT command.
I suggest:
AT+CIPSSLCERT - loads a new certificate
AT+CIPSSLCERT? - lists all certificates, each on one line
                      +CIPSSLCERT,1:DST Root CA X3 
                      +CIPSSLCERT,2:DST Root CA X3
AT+CIPSSLCERT=DELETE,1 - the comma seems to me more logical than the colon. The answer:
                     +CIPSSLCERT,1:deleted

The changes in cert deletion break an old code, we have to be aware of it.
What do you think?

jandrassy commented:
if it is possible please implement commands as documented in the "ESP8266 Non-OS AT Instruction Set" PDF.
https://www.espressif.com/sites/default/files/documentation/4a-esp8266_at_instruction_set_en.pdf

SamGeens commented:
> @SamGeens , I really appreciate your work but the fact that your changed the order of functions in command.cpp makes it nearly impossible to review and comment. Would it be much work for you to edit the file and restore the original order?
> I understand the reason why you did it and we can do it again in the future but now it's a pain in the ass to compare the original file and the new one. Sorry :(

My aplogies again for making you review so much code changes. I changed the command order to make it easier to understand and to see which command are missing and can be implemented in the future. My upcoming PR request (CWLAP, Enterprise wifi, AP modes, ...) are now already based on these changes so compelety discarting this would change a lot for me. 

But since you understand it and suggest to do it again, I think I can seperate the command order push to a new PR if you like? Which won't be that much work for me and would be much easier for you to review it. 

This would mean a few more commits since I like to keep changes seperate to easily be able to rollback if something went wrong , so definitely accept this PR as squash commit :)

SamGeens commented:
> if it is possible please implement commands as documented in the "ESP8266 Non-OS AT Instruction Set" PDF.
> https://www.espressif.com/sites/default/files/documentation/4a-esp8266_at_instruction_set_en.pdf

Ofcourse that's the one I'm following! That were the command order came from to be more unifrom with it. Everything so far is impelemented with this guide and also tested with your WiFiEspAT library.

SamGeens commented:
> I'd like to know your opinion about the syntax of AT+CIPSSLCERT command.
> I suggest:
> `AT+CIPSSLCERT` - loads a new certificate
> `AT+CIPSSLCERT?` - lists all certificates, each on one line
> 
> ```
> +CIPSSLCERT,1:DST Root CA X3 
> +CIPSSLCERT,2:DST Root CA X3
> ```
> 
> `AT+CIPSSLCERT=DELETE,1` - the comma seems to me more logical than the colon. The answer:
> 
> ```
> +CIPSSLCERT,1:deleted
> ```
> 
> The changes in cert deletion break an old code, we have to be aware of it.
> What do you think?
> 
> Edit: we can make DELETE equivalent with DELETE,1. I also expect that the sequence number of a certificate changes if a preceding one is deleted. This limits the ease of using the command.

Hi, thank you for this suggestion. 

I based the aproach I used on your previous commands but this looks more easy to understand and I can try to implement it this way.

Could you explain more deeply what breaks with the cert deletion?

JiriBilek commented:
> Could you explain more deeply what breaks with the cert deletion?

The syntax and return values. In case somebody checked it.

SamGeens commented on ESP_ATMod/command.cpp:
```diff
@@ -36,80 +36,85 @@
+	{"+CIPSSLFP", MODE_QUERY_SET, CMD_AT_CIPSSLFP},
+	{"+CIPSSLSIZE", MODE_QUERY_SET, CMD_AT_CIPSSLSIZE},
+	{"+CIPSSLCERTMAX", MODE_QUERY_SET, CMD_AT_CIPSSLCERTMAX},
+	{"+CIPSSLCERT", MODE_QUERY_SET, CMD_AT_CIPSSLCERT},
```
You're right! I'll change that

SamGeens replied:
Fixed aswell.

JiriBilek commented:
> But since you understand it and suggest to do it again, I think I can seperate the command order push to a new PR if you like? Which won't be that much work for me and would be much easier for you to review it.
> 
> This would mean a few more commits since I like to keep changes seperate to easily be able to rollback if something went wrong , so definitely accept this PR as squash commit :)

Yes, the final order could be the same as in the official documentation. I am editing the file in eclipse and this is why I don't care about the order of functions but editing in Arduino IDE and scrolling through 2000 lines of code is hard, I suppose.

So please, let us change the functionality first and after merging it we (you) can change the order of functions.

No problem with additional commits.

JiriBilek commented:
> 
> 
> if it is possible please implement commands as documented in the "ESP8266 Non-OS AT Instruction Set" PDF.
> https://www.espressif.com/sites/default/files/documentation/4a-esp8266_at_instruction_set_en.pdf

Juraj, I am aware of it. Did you find in the implementation any deviation from the documentation?

jandrassy commented:
> > if it is possible please implement commands as documented in the "ESP8266 Non-OS AT Instruction Set" PDF.
> > https://www.espressif.com/sites/default/files/documentation/4a-esp8266_at_instruction_set_en.pdf
> 
> Juraj, I am aware of it. Did you find in the implementation any deviation from the documentation?

sorry no. I only saw your discussion about the syntax and I wanted to provide a link to the pdf. also this firmware started with AT 1 commands so it would be good not to mix AT 1 and AT 2 syntax.

SamGeens commented:
Command order is already restored, except for command.h (I thought that one is not hard to read).

Next I'll implement the new command syntax  as discussed in [your comment](https://github.com/JiriBilek/ESP_ATMod/pull/4#issuecomment-881872413).

JiriBilek commented on ESP_ATMod/command.cpp:
```diff
@@ -2366,22 +2485,32 @@ commands_t findCommand(uint8_t *input, uint16_t inpLen)
 	for (unsigned int i = 0; i < sizeof(commandList) / sizeof(commandDef_t); ++i)
 	{
 		const char *cmd = commandList[i].text;
+		String compareCmd = commandList[i].text;
```
I think we should use another way to find a command. This may work but seems too complicated to me.
I'd prefer finding a token in an input string: `if (!memcmp(cmd, input + 2, strlen(cmd)))`
Then
a) MODE_EXACT_MATCH - input must continue with \r\n (i.e. check the length)
b) MODE_QUERY_SET - input must continue with either '?' or '='
c) MODE_NO_CHECKING - input must not continue with [A-Z_]

Isn't this better? BTW: you are not deleting the inputCmd buffer (memory leaks).
I can make the changes after merging the PR myself.

SamGeens replied:
Need to take a deeper look at this. The reason why I changed it was because it sometimes found the wrong command if that command was the shorter version of the command it was actually looking for.

JiriBilek replied:
Yes, I see. The commands with the same prefix were in particular order, from the longest to the shortest.

SamGeens replied:
That is ofcourse a way of fixing it, but personally I see this as a really dirty fix. If someone tries to adjust or implement something and they make the tiny mistake to just put the command somewhere they like (like I did), it can compelety mess up everything. That's why I introduced the order and a way to check commands order independent.

But I'll take a look at this and see if there is another way to do this better, but I'm quite sure this is the best I could do by checking the command on the command itself and still keeping your modechecking in place. :)

JiriBilek replied:
It was quite dirty :) but it seemed to me functional enough and (with some comment) understandable.
I think we can define the command as one of types:
- AT+COMMAND
- AT+COMMAND?
- AT+COMMAND=xxxxx

JiriBilek replied:
Oh, my bad, I forgot to check this change.
There is still the problem with memory freeing when the command is not found.
I refactor this as soon as possible, I have another idea of checking for the command.

JiriBilek commented on ESP_ATMod/command.cpp:
```diff
@@ -1742,105 +1746,68 @@ void cmd_AT_CIPSSLCERT()
+		if (!readNumber(inputBuffer, offset, certNumber) || certNumber == 0)
 		{
-			Serial.println(F("+CIPSSLCERT:no cert"));
+			Serial.printf_P(MSG_ERROR);
```
Why report error? I think it's fine to have no certificates.

SamGeens replied:
I had doubts about this one to. Have it report error or just ok, both is fine. The final thought I took is how would for example the WifiEspAt library handle this if this would be implemented, then I thought receiving error is easier than reading "no cert". 

But if you want it changed, I'll change it right now?

SamGeens replied:
The place where you pinned this comment is when the number read failed or the user gives 0, so in both cases more like an actual error than just no cert.

JiriBilek replied:
No certificate is a valid state. The libraries using ESP and this code for a wifi connection often parse OK or ERROR to know if the command failed or succeeded. I think we should return OK with no certificate loaded as I dislike "exception driven programming".

JiriBilek replied:
Sorry, the position was wrong. Added a new comment to point out the correct position.

JiriBilek commented on ESP_ATMod/ESP_ATMod.ino:
```diff
@@ -274,44 +350,42 @@ void loop()
 
 			// base 64 characters and hyphen for header and footer
-			if (isAlphaNumeric(c) || strchr("/+= -\r\n", c) != nullptr)
+			if (isAlphaNumeric(c) || strchr("/+= -\\", c) != nullptr)
```
This code does not work for me. You can' t separate and test for the backslash character as \r an \n are actually _one_ character.
Please revert to the original code.

SamGeens replied:
Okay, which certificate or layout are you using? I'm using the one I documented in the README.md and if I revert this change I get error after every line.

SamGeens replied:
The one where everything is on one line:
`-----BEGIN CERTIFICATE-----\nMIIFYDCCBEigAwIBAgIQQAF3ITfU6UK47naqPGQKtzANBgkqhkiG9w0BAQsFADA/\nMSQwIgYDVQQKExtEaWdpdGFsIFNpZ25hdHVyZSBUcnVzdCBDby4xFzAVBgNVBAMT\nDkRTVCBSb290IENBIFgzMB4XDTIxMDEyMDE5MTQwM1oXDTI0MDkzMDE4MTQwM1ow\nTzELMAkGA1UEBhMCVVMxKTAnBgNVBAoTIEludGVybmV0IFNlY3VyaXR5IFJlc2Vh\ncmNoIEdyb3VwMRUwEwYDVQQDEwxJU1JHIFJvb3QgWDEwggIiMA0GCSqGSIb3DQEB\nAQUAA4ICDwAwggIKAoICAQCt6CRz9BQ385ueK1coHIe+3LffOJCMbjzmV6B493XC\nov71am72AE8o295ohmxEk7axY/0UEmu/H9LqMZshftEzPLpI9d1537O4/xLxIZpL\nwYqGcWlKZmZsj348cL+tKSIG8+TA5oCu4kuPt5l+lAOf00eXfJlII1PoOK5PCm+D\nLtFJV4yAdLbaL9A4jXsDcCEbdfIwPPqPrt3aY6vrFk/CjhFLfs8L6P+1dy70sntK\n4EwSJQxwjQMpoOFTJOwT2e4ZvxCzSow/iaNhUd6shweU9GNx7C7ib1uYgeGJXDR5\nbHbvO5BieebbpJovJsXQEOEO3tkQjhb7t/eo98flAgeYjzYIlefiN5YNNnWe+w5y\nsR2bvAP5SQXYgd0FtCrWQemsAXaVCg/Y39W9Eh81LygXbNKYwagJZHduRze6zqxZ\nXmidf3LWicUGQSk+WT7dJvUkyRGnWqNMQB9GoZm1pzpRboY7nn1ypxIFeFntPlF4\nFQsDj43QLwWyPntKHEtzBRL8xurgUBN8Q5N0s8p0544fAQjQMNRbcTa0B7rBMDBc\nSLeCO5imfWCKoqMpgsy6vYMEG6KDA0Gh1gXxG8K28Kh8hjtGqEgqiNx2mna/H2ql\nPRmP6zjzZN7IKw0KKP/32+IVQtQi0Cdd4Xn+GOdwiK1O5tmLOsbdJ1Fu/7xk9TND\nTwIDAQABo4IBRjCCAUIwDwYDVR0TAQH/BAUwAwEB/zAOBgNVHQ8BAf8EBAMCAQYw\nSwYIKwYBBQUHAQEEPzA9MDsGCCsGAQUFBzAChi9odHRwOi8vYXBwcy5pZGVudHJ1\nc3QuY29tL3Jvb3RzL2RzdHJvb3RjYXgzLnA3YzAfBgNVHSMEGDAWgBTEp7Gkeyxx\n+tvhS5B1/8QVYIWJEDBUBgNVHSAETTBLMAgGBmeBDAECATA/BgsrBgEEAYLfEwEB\nATAwMC4GCCsGAQUFBwIBFiJodHRwOi8vY3BzLnJvb3QteDEubGV0c2VuY3J5cHQu\nb3JnMDwGA1UdHwQ1MDMwMaAvoC2GK2h0dHA6Ly9jcmwuaWRlbnRydXN0LmNvbS9E\nU1RST09UQ0FYM0NSTC5jcmwwHQYDVR0OBBYEFHm0WeZ7tuXkAXOACIjIGlj26Ztu\nMA0GCSqGSIb3DQEBCwUAA4IBAQAKcwBslm7/DlLQrt2M51oGrS+o44+/yQoDFVDC\n5WxCu2+b9LRPwkSICHXM6webFGJueN7sJ7o5XPWioW5WlHAQU7G75K/QosMrAdSW\n9MUgNTP52GE24HGNtLi1qoJFlcDyqSMo59ahy2cI2qBDLKobkx/J3vWraV0T9VuG\nWCLKTVXkcGdtwlfFRjlBz4pYg1htmf5X6DYO8A4jqv2Il9DjXA6USbW1FzXSLr9O\nhe8Y4IWS6wY7bCkjCWDcRQJMEhg76fsO3txE+FiYruq9RUWhiF1myv4Q6W+CyBFC\nDfvp7OOGAN6dEOM4+qR9sdjoSYKEBpsr6GtPAQw4dy753ec5\n-----END CERTIFICATE-----
`

JiriBilek replied:
You can't have \n as 2 characters, right? I am pasting
```
-----BEGIN CERTIFICATE-----
MIIDSjCCAjKgAwIBAgIQRK+wgNajJ7qJMDmGLvhAazANBgkqhkiG9w0BAQUFADA/	
MSQwIgYDVQQKExtEaWdpdGFsIFNpZ25hdHVyZSBUcnVzdCBDby4xFzAVBgNVBAMT	
DkRTVCBSb290IENBIFgzMB4XDTAwMDkzMDIxMTIxOVoXDTIxMDkzMDE0MDExNVow	
PzEkMCIGA1UEChMbRGlnaXRhbCBTaWduYXR1cmUgVHJ1c3QgQ28uMRcwFQYDVQQD	
Ew5EU1QgUm9vdCBDQSBYMzCCASIwDQYJKoZIhvcNAQEBBQADggEPADCCAQoCggEB	
AN+v6ZdQCINXtMxiZfaQguzH0yxrMMpb7NnDfcdAwRgUi+DoM3ZJKuM/IUmTrE4O	
rz5Iy2Xu/NMhD2XSKtkyj4zl93ewEnu1lcCJo6m67XMuegwGMoOifooUMM0RoOEq	
OLl5CjH9UL2AZd+3UWODyOKIYepLYYHsUmu5ouJLGiifSKOeDNoJjj4XLh7dIN9b	
xiqKqy69cK3FCxolkHRyxXtqqzTWMIn/5WgTe1QLyNau7Fqckh49ZLOMxt+/yUFw	
7BZy1SbsOFU5Q9D8/RhcQPGX69Wam40dutolucbY38EVAjqr2m7xPi71XAicPNaD	
aeQQmxkqtilX4+U9m5/wAl0CAwEAAaNCMEAwDwYDVR0TAQH/BAUwAwEB/zAOBgNV	
HQ8BAf8EBAMCAQYwHQYDVR0OBBYEFMSnsaR7LHH62+FLkHX/xBVghYkQMA0GCSqG	
SIb3DQEBBQUAA4IBAQCjGiybFwBcqR7uKGY3Or+Dxz9LwwmglSBd49lZRNI+DT69	
ikugdB/OEIKcdBodfpga3csTS7MgROSR6cz8faXbauX+5v3gTt23ADq1cEmv8uXr	
AvHRAosZy5Q6XkjEGB5YGV8eAlrwDPGxrancWYaLbumR9YbK+rlmM6pZW87ipxZz	
R8srzJmwN0jP41ZL9c8PDHIyh8bwRLtTcm1D9SZImlJnt1ir/md2cXjbDaJWFBM5	
JDGFoqgCWjBH4d1QB7wCCZAA62RjYJsWvIjJEubSfZGL+T0yjWW06XyxV3bqxbYo	
Ob8VZRzI9neWagqNdwvYkQsEjgfbKbYK7p2CNTUQ
-----END CERTIFICATE-----

```

Using puTTY and CtrlC/CtrlV, so that there are some end of line characters (I don't know if there is \r\n od \n).

I think that in an application it'll be something like
Serial.println("-----BEGIN CERTIFICATE-----");
Serial.println("MIIDSjCCAjKgAwIBAgIQRK+wgNajJ7qJMDmGLvhAazANBgkqhkiG9w0BAQUFADA/");
etc.

JiriBilek replied:
I hope your certificate will load fine if you delete all the "\n" strings (2 characters) in it.

JiriBilek replied:
Correction: it won't. The certificate loader requires -----BEGIN CERTIFICATE----- and -----END CERTIFICATE----- on separate lines.

JiriBilek replied:
Ok, let's combine both (not tested!):
```
			if (isAlphaNumeric(c) || strchr("/+= -\\\r\n", c) != nullptr)
			{
				// Check newlines - the header and footer must be separated by at least one '\n' from the base64 certificate data
				if (c == '\r')
					c = '\n';
				else if (c == 'n' && PemCertificatePos > 0 && PemCertificate[PemCertificatePos - 1] == '\\')
				{
					// Create real backslash from separate characters '\' and 'n'
					c = '\n';
					--PemCertificatePos;
				}

				// Ignore multiple newlines, save space in the buffer
				if (c != '\n' || (PemCertificatePos > 0 && PemCertificate[PemCertificatePos - 1] != '\n'))
				{
						PemCertificate[PemCertificatePos++] = c;

					// Check the end
					if (c == '-' && PemCertificatePos > 100)

```

SamGeens replied:
Okay, I’ll test this aswell. It seems there is quite a difference between loading with Arduino and loading with Putty

JiriBilek replied:
I have just fixed it, see above

SamGeens replied:
Yeah great it indeed fixed it! I'll push this fix then. :)

JiriBilek commented on ESP_ATMod/command.cpp:
```diff
@@ -1689,68 +1741,106 @@ void cmd_AT_CIPSSLCERT()
 		{
-			Serial.println(F("+CIPSSLCERT:no cert"));
+			Serial.println(F("+CIPSSLCERT:no certs loaded"));
+			Serial.printf_P(MSG_ERROR);
```
Sorry, the correct place for the OK message is here.

SamGeens replied:
Changed this one.

JiriBilek commented on ESP_ATMod/ESP_ATMod.h:
```diff
@@ -63,38 +63,39 @@ extern client_t clients[5];
 extern bool fingerprintValid;
-extern BearSSL::X509List *CAcert;  // CA certificate for TLS validation
+extern BearSSL::X509List CAcert; // CA certificate for TLS validation
+extern int maximumCertificates;	 // Maximum amount of certificates to load
```
should be type size_t

SamGeens replied:
Solved.

JiriBilek commented on ESP_ATMod/ESP_ATMod.ino:
```diff
@@ -76,57 +79,58 @@ const char APP_VERSION[] = "0.2.8";
 bool fingerprintValid;
-BearSSL::X509List *CAcert = nullptr;  // CA certificate for TLS validation
+BearSSL::X509List CAcert; // CA certificate for TLS validation
+int maximumCertificates;
```
should be type size_t

SamGeens replied:
Resolved.

JiriBilek commented on ESP_ATMod/command.cpp:
```diff
@@ -1689,68 +1741,106 @@ void cmd_AT_CIPSSLCERT()
 
-				if (cn != nullptr)
+				// Delete certificate
+				for (int i = 0; i < CAcert.getCount(); i++)
```
should be type size_t

SamGeens replied:
Resolved.

JiriBilek commented on ESP_ATMod/command.cpp:
```diff
@@ -1689,68 +1741,106 @@ void cmd_AT_CIPSSLCERT()
-					delete cn;
-				}
-				else
+				for (int i = 0; i < certList.getCount(); i++)
```
should be type size_t

SamGeens replied:
Resolved.

JiriBilek reviewed (commented):
The function `certList.getCount()` returns type size_t and I am getting compiler warnings about mixing signed and unsigned int.

SamGeens commented:
I pushed the certiloading fix. 

I also added 1 more boolean to check if certloading was the last action performed, because it always printed ERROR after even succesfully reading the certificate. This was due the +CIPSSLCERT command was still in the buffer, with this push that ERROR is fixes aswell.

JiriBilek commented:
> I also added 1 more boolean to check if certloading was the last action performed, because it always printed ERROR after even succesfully reading the certificate. This was due the +CIPSSLCERT command was still in the buffer, with this push that ERROR is fixes aswell.

@SamGeens I don't have the problem, I am able to paste the certificate from putty without an error.
```
AT

OK
AT+CIPSSLCERT

OK
>
Read 1216 bytes

OK
AT

OK
AT+CIPSSLCERT

OK
>
Read 1216 bytes

OK
AT+CIPSSLCERT?
+CIPSSLCERT,1:DST Root CA X3
+CIPSSLCERT,2:DST Root CA X3

OK

```

I'd not add an extra variable if it wasn't absolutely necessary, please check your code that load the certificate. I think the data (certificate) should end with '-', i.e. without a newline.

Just to mention it here, in future, we have to deal with duplicate certificates loading, but not now.

SamGeens commented:
@JiriBilek 

Could you try and do the same from the Arduino IDE Serial Monitor? (Without the extra bool added ofcourse). The problem is Arduino IDE specific I guess.

JiriBilek commented:
Using Arduino IDE Serial Monitor is tricky in this case. You have to suppress the last CRLF (after `-----END CERTIFICATE-----`). This causes the error as the single CRLF is not considered a valid command.

JiriBilek commented on ESP_ATMod/ESP_ATMod.ino:
```diff
@@ -172,6 +178,91 @@ void setup()
+			// Get filename
+			String filename = dir.fileName();
+
+			int originalCertCount = CAcert.getCount();
```
Type size_t

SamGeens replied:
Resolved.

JiriBilek commented on ESP_ATMod/ESP_ATMod.ino:
```diff
@@ -632,10 +752,46 @@ int SendData(int clientIndex, int maxSize)
+			BearSSL::X509List certList;
+
+			// Delete latest certificate
+			for (int i = 1; i < CAcert.getCount(); i++)
```
Type size_t

SamGeens replied:
Resolved.

JiriBilek commented on ESP_ATMod/ESP_ATMod.ino:
```diff
@@ -632,10 +752,46 @@ int SendData(int clientIndex, int maxSize)
+
+			CAcert = BearSSL::X509List();
+
+			for (int i = 0; i < certList.getCount(); i++)
```
Type size_t

SamGeens replied:
Resolved.

JiriBilek commented on ESP_ATMod/ESP_ATMod.ino:
```diff
@@ -632,10 +752,46 @@ int SendData(int clientIndex, int maxSize)
+	for (size_t i = 0; i < CAcert.getCount(); i++)
+	{
+		const br_x509_certificate *cert2 = &(CAcert.getX509Certs()[i]);
+		if (!memcmp(certToCheck->data, cert2->data, sizeof(certToCheck->data)) && i != indexNewAdded)
```
Is `sizeof(certToCheck->data)` correct? 
Shouldn't we test 
`certToCheck->data_len == cert2->data_len` 
and then  
`memcmp(certToCheck->data, cert2->data, certToCheck->data_len)` ?

SamGeens replied:
Resolved.

JiriBilek reviewed (commented):
Size_t instead of int and a strange test

JiriBilek commented:
And one more thought about checking for duplicates.
I'd prefer to check for duplicates _before_ adding a certificate to the list. 
Maybe it could be done with:
1. importing the certificate into another, local CAcert, 
2. check for duplicates, 
3. delete from this local CAcert (I don't know if the CAcert destruction will delete the contents) 
4. if check was ok, import the certificate into the main CAcert.

SamGeens commented:
I changed the certificate duplicate checking method and removed the extra boolean by reading everything else left on Serial after loading the certificate.

JiriBilek commented on ESP_ATMod/ESP_ATMod.ino:
```diff
@@ -324,30 +437,42 @@ void loop()
+				while (Serial.available() > 0)
+				{
+					// Read the incoming byte
+					int incomingByte = Serial.read();
```
No need to declare an unused variable. `Serial.read();` is sufficient.

SamGeens replied:
True indeed, used this earlier in my test but didn’t remove the variable. Resolved now.

JiriBilek commented on ESP_ATMod/ESP_ATMod.ino:
```diff
@@ -76,57 +80,58 @@ const char APP_VERSION[] = "0.2.8";
+int8_t gsSTNPTimezone = 0;			// command AT+CIPSNTPCFG
+String gsSNTPServer[3];				// command AT+CIPSNTPCFG
 
 /*
```
I had problems with compiling: `undeclared function checkCertificateDuplicates`. I know Arduino IDE gathers function prototypes from the .ino file (that I really hate) and maybe in the IDE it compiles fine. But my environment (Sloeber) doesn't compile. Please add the following prototype. It should not affect the Arduino IDE compilation.
```
/*
 * Local prototypes
 */
static bool checkCertificateDuplicates(BearSSL::X509List &importCertList);
```

SamGeens replied:
Ofcourse normally I add this, could not say why it isn’t there anymore because I certainly wrote this once :) Resolved aswell.

JiriBilek replied:
I'm not seeing the change. Did you commit the changes?

SamGeens replied:
Sorry, not yet commited. Busy with something now, I’m commit this in a few minutes-hour

JiriBilek replied:
no worries, I will continue tomorrow.

JiriBilek reviewed (commented):
Another comments. Please don't be mad at me ;-)

JiriBilek commented on ESP_ATMod/ESP_ATMod.ino:
```diff
@@ -172,6 +177,100 @@ void setup()
+
+						if (checkCertificateDuplicates(importCertList))
+						{
+							Serial.println(F("Tried to load already existing certificate"));
```
Missing '\n' at the start of the string.

SamGeens commented:
> Another comments. Please don't be mad at me ;-)

Haha, not mad at all. I like you commenting so much because it improves the code a lot and you see things I didn’t saw and vica versa.

JiriBilek commented:
Changes have been tested, no more issues found.
Nevertheless, I have still one request: please rename the `checkCertificateDuplicates` function as the name is misleading - it checks and loads. Rename it to `checkCertificateDuplicatesAndLoad`.
Thanks.

JiriBilek commented:
Thanks, @SamGeens for the work. All issues fixed but the one with finding a command in buffer. I will fix it soon, I don't want to torture you more :)
